Regional User Request Access

elipe17 opened the pull request:
## Summary of Changes
- Added support for acf domain based users to select whether they're central or regional users
- Added frontend support for selecting regions for regional users
- Refactored the request access form into it's own component
- Added new frontend tests to test the new form component
- Added support in the backend to allow a user to be associated with multiple regions and vice versa
- Updated email templates to handle regional vs central access request/approved emails
- Added new form fields for e2e tests
Pull request closes #3384 

## How to Test
```
cd tdrs-frontend && docker-compose up --build
cd tdrs-backend && docker-compose up --build
```

1. Open http://localhost:3000/ and sign in.
2. Request access as a normal central user, make sure you try introducing form errors to verify the errors and count of errors looks good.
3. Verify you can successfully request access and that your user is updated etc...
4. To test as the regional user with acf domain, you can modify line 247 in `RequestAccessForm` from `{isAMSUser && (` to `{!isAMSUser && (`. Notice the addition of the bang symbol.
5. Now log in, and request access again. Toggle the regional radio buttons to verify functionality.
6. Introduce/fix form errors with and without the regional checkboxes rendered
7. select one or more regions for your user, and request access
8. Verify the user is created and that it is correctly associated with the regions you chose.

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] When a user with an ACF email domain, they should be presented with the region checkboxes
- [x] Regional user is emailed when access is submitted, approved, and/or denied. 
- [x] The regional user must select 1+ region(s). 
- [x] DAC should allow regional user with null STTs if region is selected. 
- [x] Testing Checklist has been run and all tests pass
- [x] README is updated, if necessary

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

elipe17 committed:
- Moved request access form to its own component
- Pulled form elements out of Home component

elipe17 committed:
- Make many to many field for region on users model
- Create custom through model
- Update serializer to handle new many to many field
- Temporary migrations
- Update request access patch to include selected regions
- Remove unused imports
- Update regions info to be a set to ensure no duplicates and easy manipulation

elipe17 committed:
- Add inline to display new m2m field in admin

elipe17 committed:
- Updated region checkboxes to be toggleable
- Updated region checkbox error state to be smarter

elipe17 committed:
- Fixed issue where exception occurred submitting request as non regional user

elipe17 committed:
- Use correct nomenclature

elipe17 committed:
- Update email templates to reflect whether a regional or central user is submitting an access request.

elipe17 committed:
- linting

elipe17 committed:
- correct migration
- begin removing region refs

elipe17 committed:
- Updated tests to support the new regions field

elipe17 committed:
- remove select related
- remove region from factory
- update logic in tests

elipe17 committed:
- fix permission logic

elipe17 committed:
- Updated logic in validator to correctly asses if a user has a region

elipe17 committed:
- update form to display correctly for stts
- add tests for form logic

elipe17 committed:
- Added missing form fields for the regions field

elipe17 committed:
- Add todos for post review

elipe17 committed:
- revert todos because tests wont pass

elipe17 committed:
Merge branch 'develop' into 3384-regional-user-workflows

raftmsohani commented on tdrs-backend/tdpservice/users/models.py:
```diff
@@ -30,6 +30,17 @@ class AccountApprovalStatusChoices(models.TextChoices):
     # is "pending", "approved", and "denied" enough to cover functionality?
 
 
+class RegionMeta(models.Model):
```
Why do you need to add the Meta model if you are only adding the two fields for user and region?

raftmsohani replied:
I mean Django creates this automatically

elipe17 replied:
I added it just so that we didnt get into the same situation we did with the Reparse M2M stuff. I didnt want to have to accidentally create our own through model again and write a painful custom migration. Instead we have the model and if we want/need more metadata we can add it with a traditional migration.

raftmsohani commented on tdrs-backend/tdpservice/data_files/views.py:
```diff
@@ -122,6 +122,7 @@ def get_serializer_context(self):
     def download(self, request, pk=None):
         """Retrieve a file from s3 then stream it to the client."""
         record = self.get_object()
+        print(f"\n\nDOWNLOAD RECORD: {record}\n\n")
```
Is this temp?

elipe17 replied:
Definitely shouldnt be there. I don't even remember why I added that ha.

elipe17 committed:
- Make updates to allow for a11y testing

elipe17 committed:
Merge branch '3384-regional-user-workflows' of https://github.com/raft-tech/TANF-app into 3384-regional-user-workflows

raftmsohani commented:
Works as advertised!

elipe17 committed:
- Update link to correct path

elipe17 committed:
- Add aria tag
- Open link in new page
- Set form validation to OFA style val for now

elipe17 committed:
Merge branch 'develop' into 3384-regional-user-workflows

victoriaatraft commented:
<img width="1727" alt="AxeDevTools Pass" src="https://github.com/user-attachments/assets/404794d6-73f7-4b4b-9a72-624e174a7b9b" />
<img width="1725" alt="Screenshot 2025-01-29 at 12 08 45 PM" src="https://github.com/user-attachments/assets/0f6da606-c34a-4a08-8fb4-d63e587d9820" />

jtimpe commented on tdrs-backend/tdpservice/email/templates/request-approved.html:
```diff
@@ -4,7 +4,11 @@
 
-    <p>Congratulations! Your request for the TANF Data Portal (TDP) is <b>approved</b> and access as a {{ group_permission }} from {{ stt_name }} <b>has been granted</b>.</p>
+    {% if regional %}
+        <p>Congratulations! Your request for the TANF Data Portal (TDP) is <b>approved</b> and access as a {{ group_permission }} for Region(s) {{ regions }} <b>has been granted</b>.</p>
```
regions as an array comes across a little funny
<img width="373" alt="image" src="https://github.com/user-attachments/assets/46318c29-fd3b-4fdc-944c-bc8cc83b9284" />
could we use the `join` template tag filter? https://docs.djangoproject.com/en/5.1/ref/templates/builtins/#std-templatefilter-join

jtimpe commented on tdrs-backend/tdpservice/users/models.py:
```diff
@@ -205,7 +217,7 @@ def is_deactivated(self):
     def location(self):
         """Return the STT or Region based on which is not null."""
-        return self.stt if self.stt else self.region
+        return self.stt if self.stt else self.regions.all()
```
not really related to this line specifically, but there's a conflict on the data files page when logged in as a regional user
![image](https://github.com/user-attachments/assets/29c4c2ee-dd58-4d94-be58-3cdb2527bda9)
is there another ticket to address this? i think we could offer the STT combobox, but limit the stt's available to select from to only those allowed by the user's associated regions. that's too much for this ticket though

elipe17 replied:
What group did you have assigned to your user? I think you're right nonetheless and that regional users will need the box like admins but only populated with their stts. Need confirmation if this ticket is supposed to cover that or not.

elipe17 replied:
I duplicated this issue this morning. Waiting on feedback from alex wrt permissions and if the sttcombox should be implemented in this ticket.

jtimpe commented on tdrs-backend/tdpservice/email/helpers/account_status.py:
```diff
@@ -3,6 +3,14 @@
 
+def getRegionalContext(regions):
+    """Get email context for regional users."""
+    regionsNames = ['Boston', 'New York', 'Philadelphia', 'Atlanta', 'Chicago',
```
can we avoid the static list here and use the model's name attribute instead?

jtimpe replied:
also, could the line below be replaced with `__str__` or `__repr__` on the region/region-meta?

elipe17 replied:
Yes, I could use a name attribute on the Region model. Ill need to add that via a migration since the Region model only has an ID field. I will update the REPR on the Region model once I add the new name field! Good idea!

jtimpe commented on tdrs-frontend/src/components/Home/Home.jsx:
```diff
@@ -141,112 +39,7 @@ function Home() {
-        </form>
-      </div>
-    )
+    return <RequestAccessForm user={user} sttList={sttList} />
```
good move 👍

jtimpe commented on tdrs-frontend/src/components/RequestAccessForm/RequestAccessForm.jsx:
```diff
@@ -0,0 +1,358 @@
+
+    if (!Object.values(formValidation.errors).length) {
+      return dispatch(
+        requestAccess({
```
not sure if intended or not, but i can submit the form when "no" is selected for "do you work for a regional office". the resulting email is a bit confusing
<img width="718" alt="image" src="https://github.com/user-attachments/assets/c5ba8c7e-117e-4fa0-b25a-bc3e40b2b021" />

elipe17 replied:
Nice catch. I need to verify with Alex/Kathryn what the intended email is. Im not sure what ACF users were being emailed before hand since they didnt have the stt combo box either and that field is the stt name. This might be an existing thing?

elipe17 replied:
I confirmed this is an issue on develop and not an artifact of this branch. I also confirmed that `{{ group_permission }}` template value in the `request-approved.html` template is always `None` if the users group and approval are set/saved at the same time. I will be writing a ticket for that since it exists on develop too.

elipe17 replied:
New ticket: #3449

jtimpe commented on tdrs-frontend/src/components/RequestAccessForm/RequestAccessForm.jsx:
```diff
@@ -0,0 +1,358 @@
+          handleChange={handleChange}
+          handleBlur={handleBlur}
+        />
+        {isAMSUser && (
```
if you wanted to take the component-ization a little further, the three longer conditional renders look like good candidates for dumb/display components that utilize props passed in by this component. Otherwise your state handling and functions at the top level look really good!

elipe17 replied:
I thought so too, but I didn't want to go too far if it didnt seem like the right pattern to follow. I like the extra components though!

jtimpe commented on tdrs-frontend/src/components/RequestAccessForm/RequestAccessForm.jsx:
```diff
@@ -0,0 +1,358 @@
+          handleChange={handleChange}
+          handleBlur={handleBlur}
+        />
+        {isAMSUser && (
```
if you wanted to take the component-ization a little further, the three longer conditional renders look like good candidates for dumb/display components that utilize props passed in by this component. Otherwise your state handling and functions at the top level look really good!

jtimpe reviewed (changes requested):
A couple suggestions - mostly optional, but i think the email ones are the most important

elipe17 committed:
- Remove print statement
- Update email templates

elipe17 committed:
- add name field to region
- update populate stts command to use region name
- update regions csv
- add new migration for region name

elipe17 committed:
- Update to use region name

elipe17 committed:
- add name None for last region

elipe17 committed:
- Moved conditional renders into separate components
- Re-enabled expected functionality for testing

elipe17 committed:
- add key to region inputs
- move validator up in component tree
- update submit to also validate regions

elipe17 committed:
- Update migration to correctly handle region names

elipe17 committed:
- Update migration to use existing function
- Update _populate_regions function to take model object

elipe17 commented on tdrs-backend/tdpservice/stts/migrations/0011_add_region_name.py:
```diff
@@ -0,0 +1,27 @@
+DATA_DIR = Path(__file__).resolve().parent.parent / 'management/commands/data'
+def populate_regions(apps, schema_editor):
+    Region = apps.get_model('stts', 'Region')
+    _populate_regions(Region)
```
Penny for your thoughts on this. I don't like importing an app module into a migration but I didnt want to duplicate the code.

elipe17 committed:
- Remove imports and unneeded code

elipe17 committed:
- Update logic for region context

elipe17 committed:
- updated test

jtimpe commented on tdrs-backend/tdpservice/stts/migrations/0011_add_region_name.py:
```diff
@@ -0,0 +1,24 @@
+
+def populate_regions(apps, schema_editor):
+    Region = apps.get_model('stts', 'Region')
+    _populate_regions(Region)
```
getting an error on this migration 
<img width="1127" alt="image" src="https://github.com/user-attachments/assets/d660f211-3cf6-4f29-9c25-0234f9e164d2" />

elipe17 replied:
Oh yes, thats why I had it be it's own function instead of importing. I cannot get the name yet. Thank you.

elipe17 replied:
Updated the migration

elipe17 committed:
- Update migration to use custom populate function
- Revert populate stts

jtimpe reviewed (approved):
Looks great!

raftmsohani commented on tdrs-backend/tdpservice/conftest.py:
```diff
@@ -269,10 +269,10 @@ def base_regional_data_file_data(fake_file_name, regional_user):
         "section": "Active Case Data",
         "user": str(regional_user.id),
-        "region": regional_user.region.id,
+        "regions": [regional_user.regions.first().pk],
```
nit: suggest to make query specific, .first() might pass now, but might not pass later

raftmsohani commented on tdrs-backend/tdpservice/email/helpers/account_status.py:
```diff
@@ -1,6 +1,13 @@
+def getRegionalContext(regions):
+    """Get email context for regional users."""
+    return {'regional': True,
+            'regions': [str(region) for region in Region.objects.filter(id__in=regions)]}
```
you should be able to use .values_list() instead

jtimpe replied:
that wouldn't work with `str(region)` but is another good solution

raftmsohani commented on tdrs-backend/tdpservice/stts/migrations/0011_add_region_name.py:
```diff
@@ -0,0 +1,31 @@
+import csv
+from pathlib import Path
+
+def populate_regions(apps, schema_editor):
```
I strongly suggest we start using inherent django fixtures: https://docs.djangoproject.com/en/5.1/topics/db/fixtures/ instead

elipe17 replied:
Can you explain how saving the DB state to a json file and loading from that is different than loading from the already existing csv? The only real difference I see is being able to use the `django-admin` command.

elipe17 committed:
- make query a little smarter

elipe17 committed:
Merge branch 'develop' into 3384-regional-user-workflows

ADPennington committed:
Merge branch 'develop' into 3384-regional-user-workflows

ADPennington committed:
Merge branch 'develop' into 3384-regional-user-workflows

elipe17 committed:
Merge branch 'develop' into 3384-regional-user-workflows

ADPennington commented:
@elipe17 merge conflict here

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 3384-regional-user-workflows

elipe17 committed:
Merge branch '3384-regional-user-workflows' of https://github.com/raft-tech/TANF-app into 3384-regional-user-workflows

elipe17 commented:
> @elipe17 merge conflict here

@ADPennington , just fixed :)

elipe17 committed:
- Fixed migratin conflict

elipe17 committed:
- Quick logic add for null stt

ADPennington commented:
@elipe17 @lhuxraft -- @ttran-hub and I will meet tomorrow, 2/11, to test this PR together in the QASP environment. I need to avoid deleting my user account for testing in QASP because it is linked to the vast majority of data file objects tested in this environment.

ADPennington commented:
> @elipe17 @lhuxraft -- @ttran-hub and I will meet tomorrow, 2/11, to test this PR together in the QASP environment. I need to avoid deleting my user account for testing in QASP because it is linked to the vast majority of data file objects tested in this environment.

@elipe17 @lhuxraft -- @ttran-hub and I completed our testing and below are the findings. Please note that I deleted the test account after every test to model a new user flow as closely as possible. Let us know how best to proceed here. 

In summary:

- the  state issue underlying the #3449 bug is still apparent and would need to be addressed before we can onboard any regional staff users. We would be okay approving this as long as it is addressed before onboarding. 
- we also noticed that it is possible to add regions to a user account but not possible to remove any. This probably needs to be addressed in another ticket. 
- For the approved regional user's profile, would be a nice enhancement to add the region(s) associated with their account to the profile. 
- STT(s) not able to be selected on data files page (#3385)

----

**test 1. Request access as a central office user**  

user received the correct email language
![test1_email](https://github.com/user-attachments/assets/fb712d89-b360-4efa-b7f4-91caf78fd6d9)


**test 2. Request access as a regional office user (one region only)**  

user received the _incorrect_ email language (no evidence of requesting R1 in email but evident in user account)
![test2_email_and_account_LI](https://github.com/user-attachments/assets/87face99-4a92-4d8c-b535-b775cf1150cd)

**test 3. Request access as a regional office user (more than one region only)**  

user received the _incorrect_ email language (no evidence of requesting multiple regions in email but evident in user account)
![test3_email](https://github.com/user-attachments/assets/7b9077d0-3c05-492e-b321-4f67d391a8cd)

**user profile once approved**
![profile_regional](https://github.com/user-attachments/assets/38526c37-c456-4999-84a6-c3a59ad897eb)

**data files page once approved**
![data_files_regional](https://github.com/user-attachments/assets/8f35bca0-a7ed-417a-91ca-e1f9637cfdd6)

elipe17 committed:
- Update to allow admin to add or remove regions

elipe17 committed:
- Update approve template to be a little better
- Update serializer to pass M2M fields to model save to use for email context

elipe17 committed:
Merge branch 'develop' into 3384-regional-user-workflows

elipe17 committed:
- fix lint

elipe17 committed:
Merge branch '3384-regional-user-workflows' of https://github.com/raft-tech/TANF-app into 3384-regional-user-workflows

elipe17 commented:
> > @elipe17 @lhuxraft -- @ttran-hub and I will meet tomorrow, 2/11, to test this PR together in the QASP environment. I need to avoid deleting my user account for testing in QASP because it is linked to the vast majority of data file objects tested in this environment.
> 
> @elipe17 @lhuxraft -- @ttran-hub and I completed our testing and below are the findings. Please note that I deleted the test account after every test to model a new user flow as closely as possible. Let us know how best to proceed here.
> 
> In summary:
> 
> * the  state issue underlying the [Request Approved Template has Missing Data #3449](https://github.com/raft-tech/TANF-app/issues/3449) bug is still apparent and would need to be addressed before we can onboard any regional staff users. We would be okay approving this as long as it is addressed before onboarding.
> * we also noticed that it is possible to add regions to a user account but not possible to remove any. This probably needs to be addressed in another ticket.
> * For the approved regional user's profile, would be a nice enhancement to add the region(s) associated with their account to the profile.
> * STT(s) not able to be selected on data files page ([Update Data Files page for regional staff #3385](https://github.com/raft-tech/TANF-app/issues/3385))
> 
> **test 1. Request access as a central office user**
> 
> user received the correct email language ![test1_email](https://private-user-images.githubusercontent.com/63075587/412181296-fb712d89-b360-4efa-b7f4-91caf78fd6d9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk0NzIxNjksIm5iZiI6MTczOTQ3MTg2OSwicGF0aCI6Ii82MzA3NTU4Ny80MTIxODEyOTYtZmI3MTJkODktYjM2MC00ZWZhLWI3ZjQtOTFjYWY3OGZkNmQ5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjEzVDE4Mzc0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNiNDdlMGEyZWRhZTc2OTc3M2FmZjRlODg1OGQ4MTg5YzhmY2E4ZDVkOWQwNTdiMjNkOTg4MDMwNWViZmY3ZGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Z-CBfEMgBL8Rmv3ZpKl0jjZ_1R3yGOOExenVwe8WgmY)
> 
> **test 2. Request access as a regional office user (one region only)**
> 
> user received the _incorrect_ email language (no evidence of requesting R1 in email but evident in user account) ![test2_email_and_account_LI](https://private-user-images.githubusercontent.com/63075587/412182514-87face99-4a92-4d8c-b535-b775cf1150cd.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk0NzIxNjksIm5iZiI6MTczOTQ3MTg2OSwicGF0aCI6Ii82MzA3NTU4Ny80MTIxODI1MTQtODdmYWNlOTktNGE5Mi00ZDhjLWI1MzUtYjc3NWNmMTE1MGNkLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjEzVDE4Mzc0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNlNjc5MWM5MGI3NDk2ZTllOGUxNDBlOTM4N2M5MDA1NDM1ZTBiNjIxNTRiYTYxNDk4NTE0MGQwYjFmODAwOGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Ud5ZGdmdAdrJAiP_24NO9sTjSDAAO2IFY2yKSzAzvII)
> 
> **test 3. Request access as a regional office user (more than one region only)**
> 
> user received the _incorrect_ email language (no evidence of requesting multiple regions in email but evident in user account) ![test3_email](https://private-user-images.githubusercontent.com/63075587/412182889-7b9077d0-3c05-492e-b321-4f67d391a8cd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk0NzIxNjksIm5iZiI6MTczOTQ3MTg2OSwicGF0aCI6Ii82MzA3NTU4Ny80MTIxODI4ODktN2I5MDc3ZDAtM2MwNS00OTJlLWIzMjEtNGY2N2QzOTFhOGNkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjEzVDE4Mzc0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQwMDQxNGU1Mzc5NTcxNTYwZGIwODk2M2IyOTYxMzhlOTFmMWE0MmQxNzkzMjhlOGVkODI4ZmFkMTgzY2E5NDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0._W4uvz1IGPptoVQgE_uVQj4LBbzSSIT_nh5eifwyt0I)
> 
> **user profile once approved** ![profile_regional](https://private-user-images.githubusercontent.com/63075587/412184526-38526c37-c456-4999-84a6-c3a59ad897eb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk0NzIxNjksIm5iZiI6MTczOTQ3MTg2OSwicGF0aCI6Ii82MzA3NTU4Ny80MTIxODQ1MjYtMzg1MjZjMzctYzQ1Ni00OTk5LTg0YTYtYzNhNTlhZDg5N2ViLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjEzVDE4Mzc0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRiMDgwZTFkOTBiNGY2MDI4Mzc0NTM1NDU2ZGExZTJjOGY5YjQ1MWIzNDVhM2RjZmVlMzM5Yjg3OTc0ZjMxMmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.LZDce0vbg_A9ypsqBSq3D8w4aA3fs2BU7Gp4lEJ8BbY)
> 
> **data files page once approved** ![data_files_regional](https://private-user-images.githubusercontent.com/63075587/412184531-8f35bca0-a7ed-417a-91ca-e1f9637cfdd6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk0NzIxNjksIm5iZiI6MTczOTQ3MTg2OSwicGF0aCI6Ii82MzA3NTU4Ny80MTIxODQ1MzEtOGYzNWJjYTAtYTdlZC00MTdhLTkxY2EtZTFmOTYzN2NmZGQ2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjEzVDE4Mzc0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZlN2M5NzI4ZWViODY2OTkxNmY2NjE3ZDdiMGJjZTA5OTdjODVjYjY2ZDg4NzliNjUxN2M5MzZjZjYxODE1NzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.6bG2XN85ydj2Bs7_Co0ha5VnUb4Tr1duayyajZ4Jyns)

@ADPennington, I added a fix to correctly handle the email context wrt a user being central/regional. You should now be able to request access as central/regional user and get the correct emails. You can also use the same user now without having to delete it from `shell_plus`. The caveat to this is you should NOT assign the user a group while switching the request access state back and forth. It's hard to explain here why that can cause issues, but happy to get on a call to explain. The issue WRT the Group being null in approval emails is still an issue and is the same issue I fixed for the email context stuff but in a different code flow. Because it has a different flow, my fix for the emails does not apply to the group stuff. I also updated the admin form to allow you and other admins the ability to add/remove regions. Sorry this took so long, it was a tough issue to track down and has to do with Many to Many fields inside of Models.

cc. @lhuxraft

ADPennington commented:
@elipe17 re-tested with a team member who didnt previously have an account in  qasp environment. Results below:

1.   central office user request ✔️ 

![test_1_ca_central](https://github.com/user-attachments/assets/33985069-0f13-4dab-81f1-980800f55e59)

2. (after resetting account to initial state) user request for one region ✔️ 

![test_1_ca_r4_email_LI](https://github.com/user-attachments/assets/0d7799c3-ebd3-4a6e-92ca-718dcc0cf6bc)
![test_1_ca_r4_account_ar](https://github.com/user-attachments/assets/ec00ee46-8566-4281-92c9-f9be7f2685a8)

3.  (after assigning to RO group ---> account status pending ---> save/continue --> account status approval) approval email ✔️ 

![test_1_ca_r4_email_approval](https://github.com/user-attachments/assets/e00541ab-0c14-44e5-be9d-edf64c67c6e6)

4. (after adding a region to the user's approved account --> account status pending ---> save/continue --> account status approval) approval email ✔️ 
![test_1_ca_r4r1_email_LI](https://github.com/user-attachments/assets/13f8746a-051d-4347-a9bb-6ce9f1e4ab20)

5. (after checking to delete regions, removing user from RO group, and setting account status to initial state) user request for multiple regions ✔️ 
![test_1_ca_r435_account_ar](https://github.com/user-attachments/assets/992e977a-edbd-45b2-9eab-725bb73aae34)
![Screenshot 2025-02-14 111624_LI](https://github.com/user-attachments/assets/410cd86c-4f2b-4f3b-a09f-53a09a7682e8)
![Screenshot 2025-02-14 111728_LI](https://github.com/user-attachments/assets/91cb2802-a0e4-4d6f-9914-ae5997ebbf07)

ADPennington commented on tdrs-backend/tdpservice/users/test/test_api/test_users.py:
```diff
@@ -52,7 +52,7 @@ def test_create_user_endpoint_not_present(api_client, user_data):
 def test_regional_user_update_user_in_region(api_client, regional_user, user_in_region):
     """Test regional staff can update users in their region."""
-    assert regional_user.region.id == user_in_region.stt.region.id
+    assert user_in_region.stt.region in regional_user.regions.all()
```
during review, I noticed the tests here for regional staff interactions with the user accounts in/out of their region. Its come to my attention that regional user group is currently configured to have visibility into user accounts, and its not just users in their region. These permissions need to be removed before onboarding of regional staff users. Currently no impact. will spin up new ticket, as these permissions were pre-existing  @elipe17 @lhuxraft

ADPennington reviewed (approved):
LGTM @elipe17 🚀 

As discussed, there will be new tickets for:

- adding region(s) to profile for approved regional staff users
- resolving regional staff permissions re: users before onboarding
